i3lock.c

```diff
@@ -645,6 +645,120 @@ void handle_screen_resize(void) {
     redraw_screen();
 }
 
+static int read_raw_image_native(uint32_t *dest, FILE *src, int width, int height, int stride) {
+    int count = 0;
+    for (int y = 0; y < height; y++) {
+        size_t n = fread(&dest[y * stride], 1, width * 4, src);
+        count += n;
+        if (n < (size_t)(width * 4))
+            break;
+    }
+
+    return count;
+}
+
+static int read_raw_image_rgb(uint32_t *dest, FILE *src, int width, int height, int stride) {
+    unsigned char *buf = malloc(width * 3);
+    if (buf == NULL)
+        return -1;
+
+    int count = 0;
+    for (int y = 0; y < height; y++) {
+        size_t n = fread(buf, 1, width * 3, src);
+        count += n;
+        if (n < (size_t)(width * 3))
+            break;
+
+        for (int x = 0; x < width; ++x) {
+            int idx = x * 3;
+            dest[y * stride + x] = 0 |
+                                   (buf[idx + 0] << 16) |
+                                   (buf[idx + 1] << 8) |
+                                   (buf[idx + 2]);
```

Comment:
Why are "native" images in native endian, but "rgb" images are always in little endian (well, or rather "byte by byte" since 24 bits is a weird word size)? That is a bit surprising to me (but I did not read the man page diff carefully and it is documented as such).

I am not really sure what the right course of action is here, but I bet that someone will get the endianess with "native" wrong. How about making the two "channel orders" the same on big endian instead of little endian, so that people on little endian notice that the order is off and *hopefully* do the right thing? Or... how about basing both formats on bytes instead of the machine word like cairo does?

Reply:
The idea is that "native" is as close to what cairo expects as possible, to achieve the lowest overhead possible (I would probably have just used one big fread() and made the stride part of the public interface, if how cairo calculates the stride is part of its public interface, but cairo doesn't document how it calculates strides afaik, and I didn't want to rely on an implementation detail), and is meant to be used by tools; "rgb" on the other hand is there just as a convenience because RGB, with the red byte first, green byte second, and blue byte last, is the most common user-facing raw format in my experience. For example, imagemagick can output the kind of RGB this patch implements, but you can't exactly ask it for 32-bit pixels in native endianness where the highest 8 bits are ignored.

Any change would, in my opinion, have to preserve both the "native" format (for dedicated tools) and the "rgb" format (for user-created scripts), so in my opinion, the options are leaving it as it is, or adding a native-endian rgb24 version (and possibly a big-endian rgb32 version?).

Reply:
> I would probably have just used one big fread() and made the stride part of the public interface, if how cairo calculates the stride is part of its public interface, but cairo doesn't document how it calculates strides afaik, and I didn't want to rely on an implementation detail)

I would expect lots of breakage if cairo changes it requirements here, so personally, I would be fine with assuming the current stride logic to stay the way it is. For RGB24 and ARGB32, the stride logic is: There is no padding. That's especially nice to implement. ;-)

(The technical definition is: Let `bpp` be the bits per pixel of the selected format. The stride is `ceil(bpp*width/8)` rounded up to the next multiple of four. For RGB24 and ARGB32, `bpp=32` and the formula simplifies to `width*4`, i.e. just four bytes per pixel. This aligned-to-multiple-of-four is required by pixman to be able to use e.g. SSE.)

Reply:
I still think it's okay to just deal with stride as the code currently does; it doesn't hurt to be on the safe side in case cairo does actually change their stride calculations, and it shouldn't really have an effect on performance.

Can this conversation be resolved, or do you disagree?